[1단계 - 출석 구현] 차니(이동찬) 미션 제출합니다.

DongchannN opened the pull request:
<!-- 

## 코드 리뷰 팁

- 코드와 관련된 질문이 있다면, PR 본문에 적기 보다는 해당 코드를 선택하고 코멘트를 남겨주세요.
  - [참고: Adding comments to a pull request](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/reviewing-changes-in-pull-requests/commenting-on-a-pull-request#adding-comments-to-a-pull-request)

-->

## 체크 리스트

- [x] 미션의 필수 요구사항을 모두 구현했나요?
- [x] Gradle `test`를 실행했을 때, 모든 테스트가 정상적으로 통과했나요?
- [x] 애플리케이션이 정상적으로 실행되나요?
- [x] [프롤로그](https://prolog.techcourse.co.kr)에 셀프 체크를 작성했나요?
  - [프롤로그 링크](https://prolog.techcourse.co.kr/studylogs/3941)


## 객체지향 생활체조 요구사항을 얼마나 잘 충족했다고 생각하시나요?

### 1~5점 중에서 선택해주세요.

- [ ] 1 (전혀 충족하지 못함)
- [ ] 2
- [ ] 3 (보통)
- [x] 4
- [ ] 5 (완벽하게 충족)

### 선택한 점수의 이유를 적어주세요.

getter의 사용을 최소화하였으며 대부분의 값을 포장하여 각자 역할이 있는 객체로 만들었습니다.
하지만 객체들의 역할이 잘 분리되었는지 의문이 드는 점, 객체들끼리 던지는 메시지가 복잡하게 꼬여있다고 생각합니다.
또한 While문 안에 Try-Catch문이 있는 등 들여쓰기가 한 단계가 넘는 부분이 존재합니다.

  
## 어떤 부분에 집중하여 리뷰해야 할까요?

 List<Attendance>를 필드로 가지고 있는 Attendances 모델에서부터 필요한 정보나 로직이 있다면 객체를 계속 파고 들어가서 가장 내부에 있는 LocalDateTime을 관리하는 CheckInTime까지 메시지가 전달이 됩니다.
깊이가 계속 들어가다보니 구현할 때 헷갈리는 부분도 많고 모델의 역할에 혼란이 생기는 경우가 상당히 많았습니다.
그렇기에 각 모델의 역할이 충실한지, 모델에 잘못된 역할이 있는지 궁금합니다.

 또한 이번 미션의 경우에는 domain을 작성할 때만 TDD를 진행하였습니다. red는 `test`, green은 `feat`, refactor는 `refactor`로 커밋 타입을 작성하였으니 TDD에 대한 리뷰도 주시면 감사합니다 👍

kysub99 committed:
docs(README.md): 기능 구현 목록 작성

kysub99 committed:
test(CrewTest): Crew 도메인 test 코드 작성

kysub99 committed:
feat(Crew): Crew 도메인 구현

kysub99 committed:
refactor(Crew): Crew 도메인 메서드 분리

kysub99 committed:
test(CheckInTime): CheckInTime 도메인 테스트 코드 작성

kysub99 committed:
feat(CheckInTime): CheckInTime 도메인 구현

CheckInTime 에 필요한 AttendanceStatus, WorkingTime 구현

kysub99 committed:
refactor(CheckInTime): 메서드 분리 및 로직 구체화

kysub99 committed:
test(CheckInTimesTest): CheckInTimes 도메인 테스트 코드 작성

kysub99 committed:
feat(CheckInTimesTest): CheckInTimes 도메인 구현

kysub99 committed:
refactor(CheckInTimes): 중복되는 로직 메서드로 분리

kysub99 committed:
test(CheckInTimesTest): 출근 시간 추가, 수정 및 읽기에 대한 테스트 코드 작성

kysub99 committed:
feat(CheckInTimes): 출근 시간 추가, 수정 및 읽기 로직 구현

kysub99 committed:
refactor(CheckInTimes): 유효성 검사 메서드 분리 및 `stream()` 사용

kysub99 committed:
test(AttendanceTest): Attendance 도메인 테스트 코드 작성

kysub99 committed:
feat(Attendance): Attendance 도메인 로직 구현

kysub99 committed:
test(PenaltyStatusTest): PenaltyStatus 도메인 테스트 작성

kysub99 committed:
feat(PenaltyStatus): PenaltyStatus 도메인 로직 구현

kysub99 committed:
refactor(PenaltyStatus): 상수 분리 및 멤버변수 이용

kysub99 committed:
test(AttendancesTest): 출석부 집합에 대한 도메인 로직 작성

kysub99 committed:
feat(Attendances): 출석부 집합에 대한 도메인 로직 구현

kysub99 committed:
refactor(Attendances): stream()을 사용하여 가독성 향상

kysub99 committed:
test(AttendanceTest): 크루원이 출석하면 시간을 추가하는 메서드에 대한 테스트 코드 작성

kysub99 committed:
feat(Attendance): 크루원이 출석하면 시간을 추가

kysub99 committed:
feat(CsvParser): csv 파일 읽는 기능 구현

kysub99 committed:
refactor: 출석 기록 반환 시 LocalDateTime으로 변환

kysub99 committed:
refactor: 지각 상태 판단 로직 AttendanceStatus로 분리

kysub99 committed:
feat(Attendance): 특정 크루의 출석 확인 로직 구현

kysub99 committed:
feat: 출석 확인 관련 뷰 및 로직 구현

kysub99 committed:
feat: 출석 기록 수정 관련 뷰 및 로직 구현

kysub99 committed:
feat: 출석 기록 확인 관련 뷰 및 로직 구현

kysub99 committed:
feat: 면담 대상자 확인 관련 뷰 및 로직 구현

kysub99 committed:
refactor: 출석 확인 시 출력 포맷 변경

kysub99 committed:
refactor: 출석 기록 수정 시 출력 포맷 변경

kysub99 committed:
refactor: 출석 기록 확인 시 출력 포맷 변경

kysub99 committed:
refactor: 출석 대상자 확인 시 출력 포맷 변경

kysub99 committed:
refactor: 컨트롤러 메서드 분리

kysub99 committed:
refactor: 제적 대상자 확인 과정에서 상수 사용

Hyunta commented on src/main/java/domain/Crew.java:
```diff
@@ -0,0 +1,20 @@
+    public static Crew of(String name) {
+        return new Crew(name);
+    }
+}
```
TDD 플로우를 잘 지켜가면서 개발해주셨네요.
테스트와 기능 구현의 커밋을 나눠주셔서 어떻게 진행하신지 파악하기 좋았습니다.
TDD를 해보시니 어떠셨나요?

DongchannN replied:
모델에 대한 역할과 기능을 미리 생각을 하고 테스트 코드를 작성해야 하다보니 생각하는 시간이 많아지고 설계에 어려움이 있었습니다.

모델을 작성하고 바로 단위 테스트를 작성하나 테스트 코드를 작성하고 모델을 구현하나 코드를 리펙토링 할 때 TDD만의 장점을 딱히 느끼기 어려웠어요. 어짜피 리펙토링 할 때는 결국 모델의 기능들에 대한 테스트 코드가 다 작성되어 있는 것을 결과적으로 같으니까요.

TDD의 장점인 빠른 피드백도 어짜피 모델 구현 후 테스트 코드 작성 이후 피드백을 받을 텐데 그 시간을 조금 더 당긴다고 의미있을까 라는 생각도 들고요.

아직 저는 TDD 입문 단계여서 잘 못느꼈지만, 아서가 실무에서 TDD를 하실 때 느낀 장점에 대해서 궁금해요.

Hyunta replied:
TDD의 가장 큰 장점은 요구사항에 집중할 수 있다는 점에 있다고 생각합니다.
요구사항을 미리 만들어놓기 때문에 여러 기능을 동시에 개발해야할 때 누락하지 않을 수 있죠.
그리고 지나친 설계를 하지 않는 것에서 가장 큰 장점이 있다고 생각합니다.

그외에 말씀해주신 단점들 또한 명확하기 때문에 모든 경우에 대해서 잘 사용할 수 있는 방법론인지는 저도 의문입니다.
그래도 꽤나 널리 퍼진 방법론이기 때문에 이번에 경험해보시고 왜 안좋은지 명확하게 설명할 수 있게 되는 것 또한 좋은 경험이 될 것 이라고 생각합니다.

Hyunta commented on src/main/java/domain/AttendanceStatus.java:
```diff
@@ -0,0 +1,8 @@
+    LATE,
+    ABSENCE,
+    ;
+}
```
필요한 부분만 구현하신 점이 좋네요 👍🏻

Hyunta commented on src/test/java/domain/CheckInTimesTest.java:
```diff
@@ -0,0 +1,47 @@
+        CheckInTimes checkInTimes = createCheckInTimes();
+        int count = checkInTimes.countPresence();
+        assertThat(count).isEqualTo(3);
+    }
```
전반적으로 코드를 간결하게 짜주셔서 읽기 편하네요 👍🏻

DongchannN replied:
감사합니다 👍

Hyunta commented on src/main/java/controller/AttendanceController.java:
```diff
@@ -0,0 +1,117 @@
+    private final InputView inputView;
+    private final OutputView outputView;
+
+    public AttendanceController(InputView inputView,
+                                OutputView outputView) {
+        this.inputView = inputView;
+        this.outputView = outputView;
+    }
```
생성자의 파라미터는 개행하시는 컨벤션을 사용하시는 걸까요?
그렇다면 Attendance의 생성자 컨벤션도 통일해주세요.

```java

    private Attendance(Crew crew, CheckInTimes checkInTimes) {
        this.crew = crew;
        this.checkInTimes = checkInTimes;
    }
```

```java

    WorkingTime(int startHour,
                int startMinute,
                int endHour,
                int endMinute) {
        this.startHour = startHour;
        this.startMinute = startMinute;
        this.endHour = endHour;
        this.endMinute = endMinute;
    }
```

DongchannN replied:
통일되지 못한 부분이 있네요. 다음 리뷰에 피드백 적용하겠습니다!

Hyunta commented on src/main/java/util/AttendanceParser.java:
```diff
@@ -0,0 +1,43 @@
+package util;
+
+import domain.Attendance;
+import domain.Attendances;
+import domain.CheckInTimes;
+import domain.Crew;
+
+import java.time.LocalDateTime;
+import java.time.format.DateTimeFormatter;
+import java.util.ArrayList;
+import java.util.List;
+
+public class AttendanceParser {
```
`AttendanceParser` 클래스를 util 하위 패키지에 위치시키신 이유가 있을까요?
util 패키지 하위에는 어떤 클래스가 들어가면 좋을까요?

DongchannN replied:
`AttendanceParser`의 위치에 대해서 3가지를 고민해봤어요.

1. `Attendances` domain에 넣는다.
파싱의 반환값이 `Attendances` 모델이기 때문에 모델의 메서드로 넣을 수 있겠다는 생각이 들었어요. 하지만 file의 입출력과 관련되어 있고 해당 파일이 csv인지, json인지 등 포맷이 달라질 수 있다는 점을 고려해서 파일의 포멧에 따라 구현이 달라질 수 있기에 `domain`과는 분리가 되어야한다고 생각했습니다.

2. Controller의 메서드로 넣는다.
파싱하는 메서드가 `Controller`의 역할이 되기에 무리가 없다고 생각해요. 하지만 지금 컨트롤러에서 기능의 번호를 입력받고 적합한 메서드를 호출하는 등 역할이 많다고 생각했습니다.

3. util로 넣는다.
비즈니스적인 로직이 없고 단순히 모델로 변환하는 역할만 가지고 있기 때문에 `controller`와 분리하는 것이 가장 좋다고 생각해서 일단 util 패키지 안에 넣었습니다. 만약 `csv` 뿐만 아니라 다른 파일 포맷도 생긴다면 파싱하는 부분은 `domain`, `controller`와 분리되는 것이 맞다고 생각했어요. 추상화를 시켜서 해당 파일이 어떤 형식인지 상관하지 않고 `controller`의 입장해서는 `Attendacnes` 모델을 받을 수 있고요.

파싱하는 부분의 위치에 대한 아서의 생각도 궁금합니다. 위 세가지 위치 말고 다른 방안이 있을지도 알려주세요!

Hyunta replied:
여러 방면으로 적절하게 고민을 많이 해주셨네요.
코드를 간결하게 짜시는 것부터 고민도 명확하게 해주시네요 👍🏻

저는 파싱하는 부분과 파싱된 값을 객체로 만드는 부분을 분리해서 관리할 것 같습니다.
1번에서 말씀해주신 것처럼 parsing 기능은 파일 입출력과 관련있기 때문에 도메인에 넣기에는 어색합니다.
그래서 해당 부분을 분리해서 원시값을 통해서 객체를 만드는 부분과 별도로 관리하도록 구현할 것 같습니다.

DongchannN replied:
네 한번 고민해서 2단계에 적절하게 피드백 적용해볼게요!

Hyunta commented on src/main/java/util/AttendanceParser.java:
```diff
@@ -0,0 +1,43 @@
+        List<List<String>> rawAttendances = CsvParser.readFile(fileName);
+
+        List<Attendance> attendances = new ArrayList<>();
+
+        Attendances entity = Attendances.of(attendances);
```
변수명이 entity 보다는 구체적이었으면 좋겠네요

Hyunta commented on src/main/java/util/AttendanceParser.java:
```diff
@@ -0,0 +1,43 @@
+
+        Attendances entity = Attendances.of(attendances);
+
+        for (List<String> line : rawAttendances) {
+            String crewName = line.get(0);
+            try {
+                Attendance attendanceByName = entity.findAttendanceByName(crewName);
+                LocalDateTime time = getLocalDateTime(line.get(1), formatter);
+                attendanceByName.checkIn(time);
+            } catch (IllegalArgumentException e) {
+                Crew crew = Crew.of(crewName);
+                Attendance newAttendance = Attendance.of(crew, CheckInTimes.of(List.of()));
+                LocalDateTime time = getLocalDateTime(line.get(1), formatter);
+                newAttendance.checkIn(time);
+                attendances.add(newAttendance);
+            }
+        }
```
출석 명단에서 해당 이름이 존재하지 않을 경우 예외를 통해 분기처리를 하도록 설계되어있네요.
만약 이름은 존재하는데 다른 이유로 IllegalArgumentException이 발생하면 어떤 문제가 발생할까요?

동료 개발자들이 해당 코드를 봤을 때 분기처리가 아니라 예외처리로 작성되어있어서 헷갈릴 수 있을 것 같아요.
현재는 catch문에서 처리하는 로직이 예외가 발생한 상황이 아니라 정상적인 분기처리로 판단되서요.

DongchannN replied:
다양한 문제가 있을 경우 `IllegalArgumentException`으로 예외처리를 했을 때, 하나의 경우만 고려해서 로직을 수행하면 제가 원하지 않을 상황에서 해당 로직이 수행될 수 있겠네요.

`Exception` 발생 유무로 분기를 판단하는 것이 아니라 해당 조건이 분기를 일으킬지 정확히 판단하고 로직을 수행하는 것이 맞겠네요!
좋은 의견 감사합니다 👍

Hyunta commented on src/main/java/controller/AttendanceController.java:
```diff
@@ -0,0 +1,117 @@
+        }
+    }
+
+    private void selectFeature(Attendances attendances, String featureNumber) {
+        if (featureNumber.equals("1")) {
+            checkIn(attendances);
+            return;
+        }
+        if (featureNumber.equals("2")) {
+            modifyCheckInTime(attendances);
+            return;
+        }
+        if (featureNumber.equals("3")) {
+            readCheckInTime(attendances);
+            return;
+        }
+        if (featureNumber.equals("4")) {
+            readDangerCrews(attendances);
+            return;
+        }
+        throw new IllegalArgumentException("[ERROR] 1, 2, 3, 4, Q 만 입력해주세요.");
+    }
```
직관적으로 짜주셔서 읽기 좋습니다.
혹시 featureNumber에 대해서는 enum을 사용해볼까 고민을 해보셨을까요?

DongchannN replied:
enum 사용에 대해서는 고려하지 않았습니다. 직접적으로 기능에 대한 숫자를 판단하는 것이 아닌 다른 방법을 생각해볼게요!

Hyunta commented on src/main/java/controller/AttendanceController.java:
```diff
@@ -0,0 +1,117 @@
+        throw new IllegalArgumentException("[ERROR] 1, 2, 3, 4, Q 만 입력해주세요.");
+    }
+
+    private void checkIn(Attendances attendances) {
+        String name = inputView.readNickName();
+        Attendance attendanceByName = attendances.findAttendanceByName(name);
+
+        LocalDateTime checkInTime = getCheckInTime();
+        attendanceByName.checkIn(checkInTime);
+
+        outputView.printTodayCheckInTime(CheckInTime.of(checkInTime));
+    }
```
[제안]
이름을 통해 출석자를 조회했을 때 변수명을 `attendanceByName` 로 대부분 작성해주셨습니다.
개인적으로 `attendanceByName` 에서 `ByName` 부분은 불필요한 것 같아요.
이후에 로직을 수행함에 있어서 이름을 통해 조회되었다는 내용이 중요한 부분은 아닌 것 같아서요.

Hyunta commented on src/main/java/domain/Attendances.java:
```diff
@@ -0,0 +1,39 @@
+            }
+        }
+        return dangerAttendances;
+    }
+
+
+}
```
빈 2줄은 쓸 일이 거의 없습니다.

```suggestion
    }
}
```

Hyunta commented on src/main/java/controller/AttendanceController.java:
```diff
@@ -0,0 +1,117 @@
+        this.outputView = outputView;
+    }
+
+    public void run() {
+        String filePath = "src/main/resources/attendances.csv";
+        Attendances attendances = AttendanceParser.registerAttendances(
+                filePath,
+                DateTimeFormatter.ofPattern("yyyy-MM-dd HH:mm")
+        );
+        readFeature(attendances);
+    }
```
Attandances 를 계속 파라미터로 전달하면서 사용하도록 구현해주셨는데,
인스턴스를 생성하는 시점에 인스턴스 변수로 저장해놓고 관리하면 매번 주고받지 않아도 괜찮지 않을까요?

Hyunta reviewed (changes requested):
안녕하세요 차니,
출석 미션 리뷰어로 만나게 된 아서입니다.
먼저 리뷰가 늦어진 점 사과드립니다.
코드와 설계를 모두 간결하게 가져가주셔서 읽기 편했습니다.
몇가지 커멘트 남겼는데 확인해보시고 다시 요청주세요.

DongchannN committed:
refactor: 생성자의 파라미터에 대한 Multi-line 컨벤션 통일

DongchannN committed:
style: 변수명 수정

DongchannN committed:
refactor: `findAttendanceByName()` 반환 타입을 `Optional`로 변경

기존에는 해당 이름이 없을 경우 예외를 던졌으나, 이제 `Optional<Attendance>`를 반환하여 호출하는 쪽에서 후처리를 결정하도록 수정

DongchannN committed:
style: 불필요한 개행 제거

DongchannN committed:
refactor: `Attendance attendance` 지역 변수에서 인스턴스 변수로 변경

DongchannN committed:
fix: "Q"입력 시 예외 발생 해결

DongchannN committed:
style: 디렉토리 분리

DongchannN committed:
refactor: Feature enum 도입

DongchannN committed:
style: 기능 입력 부분 메서드 통합

DongchannN commented:
## 피드백  반영 사항
- [x] Multi-line convention 통일하기
- [x] AttendanceParser -> entity 변수명 수정
- [x] AttendanceController -> attendanceByName 변수명 수정
- [x] AttendanceParser -> IllegalArgumentException 사용하는 분기 수정
- [x] featureNumber에 대한 처리 고민 -> Feature enum 도입
- [x] Attendances -> 불필요한 개행 제거
- [x] 컨트롤러의 `Attendances attendances` 인스턴스 변수로 변경

두번째 코드리뷰를 받기 전에 궁금한 점이 있습니다!

**1. 코드 스타일**
일단 ab7970d 의 코드와 6792625 의 코드 관련해서, 첫번째 코드 같은 경우는 indent가 2 이상이 되면 안된다는 요구사항을 지켰지만 하나의 행동에 메서드 분리가 너무 많아 오히려 가독성을 해치는 느낌을 받았습니다. 그래서 indent 요구사항을 포기하고 두번째 코드와 같이 다시 코드 스타일을 바꾸었는데 아서의 의견이 궁금합니다

**2. 모델의 세분화로 연쇄되는 메서드 호출**
모델의 역할을 세분화하다보니 기능을 호출할 때 점점 깊게 메서드가 호출되고 있습니다.
`Attendances`모델에서 `Attendance`모델을 추출해서 출석 시간을 수정하게 되면 `Attendance`의 메서드를 호출하고 그 메서드가 `CheckInTimes`의 메서드를 호출하고 또 `CheckInTime`메서드까지 호출하게 됩니다.

만약 모델이 더 많아지면 너무 깊어지지 않게 적당히 끊는 방법도 생각해야한다고 들었는데 구체적으로 어떻게 끊는지 궁금합니다!

Hyunta commented:
> 1. 코드 스타일
일단 https://github.com/woowacourse/java-attendance/commit/ab7970ded6ca3d9b26c3c4e6c546d6bf093d337c 의 코드와 https://github.com/woowacourse/java-attendance/commit/679262597a2c54c7074ff8d4bd33375c4deacc3f 의 코드 관련해서, 첫번째 코드 같은 경우는 indent가 2 이상이 되면 안된다는 요구사항을 지켰지만 하나의 행동에 메서드 분리가 너무 많아 오히려 가독성을 해치는 느낌을 받았습니다. 그래서 indent 요구사항을 포기하고 두번째 코드와 같이 다시 코드 스타일을 바꾸었는데 아서의 의견이 궁금합니다

네 사용자 입력받는 부분은 while문 때문에 어쩔 수 없이 depth가 많이 발생하는 부분이라고 생각합니다.
의도가 합리적이라고 생각해서 depth 2 에 대한 예외를 적용해도 괜찮다는 의견입니다.

> 2. 모델의 세분화로 연쇄되는 메서드 호출
모델의 역할을 세분화하다보니 기능을 호출할 때 점점 깊게 메서드가 호출되고 있습니다.
Attendances모델에서 Attendance모델을 추출해서 출석 시간을 수정하게 되면 Attendance의 메서드를 호출하고 그 메서드가 CheckInTimes의 메서드를 호출하고 또 CheckInTime메서드까지 호출하게 됩니다.
만약 모델이 더 많아지면 너무 깊어지지 않게 적당히 끊는 방법도 생각해야한다고 들었는데 구체적으로 어떻게 끊는지 궁금합니다!

제가 생각할 때는 세분화가 객체 지향 설계에서 가져가는 가장 큰 장점이자 단점입니다.
두가지 코드를 비교해볼게요 
A코드: Attendances, Attendance, CheckInTimes, CheckInTime 객체 분리
B코드: Attendances 객체만 존재

전체 로직을 파악하는데 있어서는 A 보다는 B가 수월할 수 있습니다.
하나의 파일안에 전체 로직이 작성되어 있으니 다른 파일을 읽어볼 필요가 전혀 없습니다.
반면에 A는 믿음으로 가야합니다. 예를 들어 Attendances에서 `Attendance.countLate()`의 메서드를 사용하고 있으면  정확히 어떻게 구현됐는지는 모르지만 지각 횟수를 세나보다 하고 넘어갑니다. 그래서 정확히 디버깅을 하려면 다른 파일도 읽어봐야 합니다.

지금은 객체지향을 학습하고 있기 때문에 객체 분리에 신경을 쓰고 바라봅니다. 그런데 항상 분리하는 것이 좋은 일만 있지는 않습니다.
말씀해주신 것처럼 때로는 합쳐져있는게 더 나을 수도 있죠. 지금과 같은 경우에서는 객체 분리를 하지 않으면 객체의 깊이가 깊어지지 않습니다.

Hyunta reviewed (approved):
안녕하세요 차니,
1단계 미션 잘 진행해주셨네요 👍🏻
커멘트에 대한 답변 달아드렸으니 확인부탁드립니다.
궁금한 점이 있으면 DM 주세요.
